File: lib/combat/WeaponDamage.js
/** Base bullet damage before zone mult and weapon profile scale. */
export const TARGET_DAMAGE = 6;

/** At full ADS, bullet damage is multiplied by this factor (linear blend from hip). */
export const ADS_DAMAGE_MULTIPLIER = 1.3;

/** At full ADS, recoil strength is multiplied by (1 - this). */
export const ADS_RECOIL_REDUCTION = 0.4;

/** @param {number} aimBlend 0 hip → 1 ADS */
export function resolveAdsDamageScale(aimBlend) {
  const t = Math.max(0, Math.min(1, aimBlend));
  return 1 + t * (ADS_DAMAGE_MULTIPLIER - 1);
}

/** @param {number} aimBlend 0 hip → 1 ADS */
export function resolveAdsRecoilScale(aimBlend) {
  const t = Math.max(0, Math.min(1, aimBlend));
  return 1 - t * ADS_RECOIL_REDUCTION;
}

/** @typedef {'rifle' | 'pistol'} WeaponDamageId */

/** Legs + pelvis + feet — pistol uses {@link WeaponDamageProfile.lowerBodyScale}. */
export const LOWER_BODY_HIT_ZONES = new Set([
  "pelvis",
  "thigh",
  "knee",
  "lower_leg",
  "foot",
]);

/**
 * Per-weapon damage against humanoid hit zones (multiplies zone `mult` × base).
 * @typedef {{
 *   id: WeaponDamageId,
 *   label: string,
 *   baseDamage: number,
 *   headshotHealthFraction: number,
 *   upperBodyScale: number,
 *   lowerBodyScale: number,
 *   armScale: number,
 * }} WeaponDamageProfile
 */

/** @type {Record<WeaponDamageId, WeaponDamageProfile>} */
export const WEAPON_DAMAGE_PROFILES = {
  rifle: {
    id: "rifle",
    label: "Rifle",
    baseDamage: TARGET_DAMAGE,
    headshotHealthFraction: 1,
    upperBodyScale: 1,
    lowerBodyScale: 1,
    armScale: 1,
  },
  pistol: {
    id: "pistol",
    label: "Pistol",
    baseDamage: TARGET_DAMAGE,
    headshotHealthFraction: 1,
    upperBodyScale: 0.5,
    lowerBodyScale: 0.25,
    armScale: 0.25,
  },
};

/** Damage multiplier by camera-to-hit distance. Tuned around 30HP targets. */
export const WEAPON_DAMAGE_FALLOFF = {
  pistol: [
    { distance: 0, multiplier: 1 },
    { distance: 3, multiplier: 1 },
    { distance: 5, multiplier: 0.5 },
    { distance: 10, multiplier: 1 / 3 },
    { distance: 50, multiplier: 0.25 },
  ],
  rifle: [
    { distance: 0, multiplier: 1 },
    { distance: 10, multiplier: 1 },
    { distance: 50, multiplier: 0.5 },
  ],
};

/** @param {string} weaponId */
export function getWeaponDamageProfile(weaponId) {
  return (
    WEAPON_DAMAGE_PROFILES[/** @type {WeaponDamageId} */ (weaponId)] ??
    WEAPON_DAMAGE_PROFILES.rifle
  );
}

/**
 * @param {WeaponDamageProfile | WeaponDamageId} profileOrId
 * @param {number} shotDistance Camera-to-hit distance in metres.
 */
export function resolveDamageFalloff(profileOrId, shotDistance = 0) {
  const id =
    typeof profileOrId === "string"
      ? profileOrId
      : profileOrId?.id ?? "rifle";
  const points =
    WEAPON_DAMAGE_FALLOFF[/** @type {WeaponDamageId} */ (id)] ??
    WEAPON_DAMAGE_FALLOFF.rifle;
  const distance = Math.max(
    0,
    Number.isFinite(shotDistance) ? shotDistance : 0
  );

  if (distance <= points[0].distance) return points[0].multiplier;
  for (let i = 1; i < points.length; i += 1) {
    const prev = points[i - 1];
    const next = points[i];
    if (distance > next.distance) continue;
    const span = Math.max(0.0001, next.distance - prev.distance);
    const t = (distance - prev.distance) / span;
    return prev.multiplier + (next.multiplier - prev.multiplier) * t;
  }
  return points[points.length - 1].multiplier;
}

/** @param {string} zoneId */
export function hitZoneBodyCategory(zoneId) {
  if (zoneId === "head") return "head";
  if (zoneId === "arm") return "arm";
  if (LOWER_BODY_HIT_ZONES.has(zoneId)) return "lower";
  return "upper";
}

/**
 * @param {WeaponDamageProfile} profile
 * @param {number} currentHealth
 * @param {number} maxHealth
 * @param {number} [shotDistance]
 */
export function resolveHeadshotDamage(
  profile,
  currentHealth,
  maxHealth,
  shotDistance = 0
) {
  const falloff = resolveDamageFalloff(profile, shotDistance);
  if (profile.headshotHealthFraction >= 1 && falloff >= 0.999) {
    return currentHealth;
  }
  return maxHealth * profile.headshotHealthFraction * falloff;
}

/**
 * @param {WeaponDamageProfile} profile
 * @param {string} zoneId
 * @param {number} zoneMult
 * @param {number} [shotDistance]
 */
export function resolveBodyZoneDamage(
  profile,
  zoneId,
  zoneMult,
  shotDistance = 0
) {
  const category = hitZoneBodyCategory(zoneId);
  let scale = profile.upperBodyScale;
  if (category === "lower") scale = profile.lowerBodyScale;
  else if (category === "arm") scale = profile.armScale;
  return (
    profile.baseDamage *
    zoneMult *
    scale *
    resolveDamageFalloff(profile, shotDistance)
  );
}

File: rust/game_core/src/weapon_damage.rs
use serde::{Deserialize, Serialize};

pub const TARGET_DAMAGE: f64 = 6.0;
pub const ADS_DAMAGE_MULTIPLIER: f64 = 1.3;
pub const ADS_RECOIL_REDUCTION: f64 = 0.4;

pub fn resolve_ads_damage_scale(aim_blend: f64) -> f64 {
    let t = aim_blend.clamp(0.0, 1.0);
    1.0 + t * (ADS_DAMAGE_MULTIPLIER - 1.0)
}

pub fn resolve_ads_recoil_scale(aim_blend: f64) -> f64 {
    let t = aim_blend.clamp(0.0, 1.0);
    1.0 - t * ADS_RECOIL_REDUCTION
}

#[derive(Clone, Copy, Debug)]
struct FalloffPoint {
    distance: f64,
    multiplier: f64,
}

#[derive(Clone, Copy, Debug)]
struct WeaponDamageProfile {
    base_damage: f64,
    headshot_health_fraction: f64,
    upper_body_scale: f64,
    lower_body_scale: f64,
    arm_scale: f64,
}

const RIFLE_PROFILE: WeaponDamageProfile = WeaponDamageProfile {
    base_damage: TARGET_DAMAGE,
    headshot_health_fraction: 1.0,
    upper_body_scale: 1.0,
    lower_body_scale: 1.0,
    arm_scale: 1.0,
};

const PISTOL_PROFILE: WeaponDamageProfile = WeaponDamageProfile {
    base_damage: TARGET_DAMAGE,
    headshot_health_fraction: 1.0,
    upper_body_scale: 0.5,
    lower_body_scale: 0.25,
    arm_scale: 0.25,
};

const PISTOL_FALLOFF: &[FalloffPoint] = &[
    FalloffPoint {
        distance: 0.0,
        multiplier: 1.0,
    },
    FalloffPoint {
        distance: 3.0,
        multiplier: 1.0,
    },
    FalloffPoint {
        distance: 5.0,
        multiplier: 0.5,
    },
    FalloffPoint {
        distance: 10.0,
        multiplier: 1.0 / 3.0,
    },
    FalloffPoint {
        distance: 50.0,
        multiplier: 0.25,
    },
];

const RIFLE_FALLOFF: &[FalloffPoint] = &[
    FalloffPoint {
        distance: 0.0,
        multiplier: 1.0,
    },
    FalloffPoint {
        distance: 10.0,
        multiplier: 1.0,
    },
    FalloffPoint {
        distance: 50.0,
        multiplier: 0.5,
    },
];

fn weapon_profile(weapon_id: &str) -> WeaponDamageProfile {
    if weapon_id == "pistol" {
        PISTOL_PROFILE
    } else {
        RIFLE_PROFILE
    }
}

pub fn resolve_damage_falloff(weapon_id: &str, shot_distance: f64) -> f64 {
    let points = if weapon_id == "pistol" {
        PISTOL_FALLOFF
    } else {
        RIFLE_FALLOFF
    };
    let distance = if shot_distance.is_finite() {
        shot_distance.max(0.0)
    } else {
        0.0
    };

    if distance <= points[0].distance {
        return points[0].multiplier;
    }
    for i in 1..points.len() {
        let prev = points[i - 1];
        let next = points[i];
        if distance > next.distance {
            continue;
        }
        let span = (next.distance - prev.distance).max(0.0001);
        let t = (distance - prev.distance) / span;
        return prev.multiplier + (next.multiplier - prev.multiplier) * t;
    }
    points[points.len() - 1].multiplier
}

fn hit_zone_body_category(zone_id: &str) -> &'static str {
    if zone_id == "head" {
        "head"
    } else if zone_id == "arm" {
        "arm"
    } else if matches!(
        zone_id,
        "pelvis" | "thigh" | "knee" | "lower_leg" | "foot"
    ) {
        "lower"
    } else {
        "upper"
    }
}

pub fn resolve_headshot_damage(
    weapon_id: &str,
    current_health: f64,
    max_health: f64,
    shot_distance: f64,
) -> f64 {
    let profile = weapon_profile(weapon_id);
    let falloff = resolve_damage_falloff(weapon_id, shot_distance);
    if profile.headshot_health_fraction >= 1.0 && falloff >= 0.999 {
        return current_health;
    }
    max_health * profile.headshot_health_fraction * falloff
}

pub fn resolve_body_zone_damage(
    weapon_id: &str,
    zone_id: &str,
    zone_mult: f64,
    shot_distance: f64,
) -> f64 {
    let profile = weapon_profile(weapon_id);
    let category = hit_zone_body_category(zone_id);
    let scale = match category {
        "lower" => profile.lower_body_scale,
        "arm" => profile.arm_scale,
        _ => profile.upper_body_scale,
    };
    profile.base_damage * zone_mult * scale * resolve_damage_falloff(weapon_id, shot_distance)
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TargetZoneDamageOutput {
    pub zone: String,
    pub damage: f64,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pistol_falloff_at_five_metres() {
        assert!((resolve_damage_falloff("pistol", 5.0) - 0.5).abs() < 0.0001);
    }

    #[test]
    fn headshot_one_shot_at_full_health() {
        let dmg = resolve_headshot_damage("rifle", 30.0, 30.0, 0.0);
        assert!((dmg - 30.0).abs() < 0.0001);
    }

    #[test]
    fn pistol_pelvis_uses_lower_body_scale() {
        let dmg = resolve_body_zone_damage("pistol", "pelvis", 1.0, 0.0);
        assert!((dmg - 1.5).abs() < 0.0001);
    }

    #[test]
    fn ads_damage_scale_at_full_ads() {
        assert!((resolve_ads_damage_scale(1.0) - ADS_DAMAGE_MULTIPLIER).abs() < 0.0001);
        assert!((resolve_ads_damage_scale(0.0) - 1.0).abs() < 0.0001);
    }

    #[test]
    fn ads_recoil_scale_at_full_ads() {
        assert!((resolve_ads_recoil_scale(1.0) - 0.6).abs() < 0.0001);
        assert!((resolve_ads_recoil_scale(0.0) - 1.0).abs() < 0.0001);
    }
}
